File: src/cpp/parallel/array_sum_parallel.cpp
#include <iostream>
#include <vector>
#include <random>
#include <omp.h>

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <array_size> <num_threads>" << std::endl;
        return 1;
    }

    size_t size = std::stoull(argv[1]);
    int num_threads = std::stoi(argv[2]);
    
    // Generate random numbers
    std::vector<int> numbers(size);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1, 100);
    
    for (size_t i = 0; i < size; ++i) {
        numbers[i] = dis(gen);
    }
    
    long long sum = 0;
    double start_time = omp_get_wtime();
    
    #pragma omp parallel num_threads(num_threads) reduction(+:sum)
    {
        #pragma omp for
        for (size_t i = 0; i < size; ++i) {
            sum += numbers[i];
        }
    }
    
    double end_time = omp_get_wtime();
    
    std::cout << sum << "," << (end_time - start_time) << std::endl;
    return 0;
} 

File: src/fortran/parallel/array_sum_parallel.f90
program array_sum_parallel
    use omp_lib
    implicit none
    
    integer :: size, num_threads, i
    integer, allocatable :: numbers(:)
    integer(kind=8) :: sum
    real(kind=8) :: start_time, end_time
    character(len=32) :: arg
    
    ! Get command line arguments
    if (command_argument_count() /= 2) then
        write(*,*) "Usage: ./array_sum_parallel <size> <num_threads>"
        stop
    endif
    
    call get_command_argument(1, arg)
    read(arg, *) size
    call get_command_argument(2, arg)
    read(arg, *) num_threads
    
    allocate(numbers(size))
    
    ! Generate random numbers
    call random_seed()
    do i = 1, size
        call random_number(numbers(i))
        numbers(i) = int(numbers(i) * 100) + 1
    end do
    
    sum = 0
    call omp_set_num_threads(num_threads)
    start_time = omp_get_wtime()
    
    !$omp parallel do reduction(+:sum)
    do i = 1, size
        sum = sum + numbers(i)
    end do
    !$omp end parallel do
    
    end_time = omp_get_wtime()
    
    write(*,*) sum, ",", end_time - start_time
    
    deallocate(numbers)
end program array_sum_parallel 